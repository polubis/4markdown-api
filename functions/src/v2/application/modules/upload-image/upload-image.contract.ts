import { base64 } from '@utils/validators';
import { z } from 'zod';

const uploadImagePayloadSchema = z.object({
  image: base64,
});

type UploadImagePayload = z.infer<typeof uploadImagePayloadSchema>;

export type { UploadImagePayload };
export { uploadImagePayloadSchema };
