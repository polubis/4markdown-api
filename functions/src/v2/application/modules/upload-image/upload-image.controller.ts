import { protectedController } from '@utils/controller';
import { parse } from '@utils/parse';
import {
  updateDocumentNamePayloadSchema,
  type UpdateDocumentNameDto,
} from './update-document-name.contract';
import { updateDocumentNameHandler } from './update-document-name.handler';
import { decodeBase64Asset } from '@libs/decoding/decode-base-64-asset';

const uploadImageController = protectedController<UpdateDocumentNameDto>(
  async (rawPayload, context) => {
    decodeBase64Asset(rawPayload)
    return updateDocumentNameHandler({
      payload: await parse(updateDocumentNamePayloadSchema, rawPayload),
      context,
    });
  },
);

export { uploadImageController };
