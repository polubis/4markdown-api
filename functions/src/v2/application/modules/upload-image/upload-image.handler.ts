import { type ProtectedControllerHandlerContext } from '@utils/controller';
import {
  type UploadImagePayload,
  type UploadImageDto,
} from './upload-image.contract';
import { decodeBase64Asset } from '@utils/decode-base64-asset';
import {
  IMAGE_CONTENT_TYPES,
  IMAGE_EXTENSIONS,
  ImageContentType,
  ImageExtension,
  ImagesModel,
} from '@domain/models/image';
import { errors } from '@utils/errors';
import { toUnit } from '@libs/helpers/to-unit';
import { storage } from 'firebase-admin';
import { uuid } from '@libs/helpers/stamps';
import { Id } from '@utils/validators';

const isSupportedContentType = (
  contentType: string,
): contentType is ImageContentType =>
  IMAGE_CONTENT_TYPES.includes(contentType as ImageContentType);

const isSupportedExtension = (extension: string): extension is ImageExtension =>
  IMAGE_EXTENSIONS.includes(extension as ImageExtension);

const decodeImage = (
  image: UploadImagePayload['image'],
): ReturnType<typeof decodeBase64Asset> => {
  try {
    return decodeBase64Asset(image);
  } catch {
    throw errors.badRequest(`Cannot decode. Invalid image format`);
  }
};

// @TODO[PRIO=2]: [Add bucket + storage to injected API's??? Or abstract it?].
const createUrl = ({
  id,
  uid,
  name,
}: {
  name: string;
  id: Id;
  uid: Id;
}): string => {
  const location = `${uid}/images/${id}`;

  return `https://firebasestorage.googleapis.com/v0/b/${name}/o/${encodeURIComponent(
    location,
  )}?alt=media`;
};

const uploadImageHandler = async ({
  payload,
  context: { db, uid },
}: {
  payload: UploadImagePayload;
  context: ProtectedControllerHandlerContext;
}): Promise<UploadImageDto> => {
  const { contentType, extension, size, buffer } = decodeImage(payload.image);

  if (!isSupportedContentType(contentType)) {
    throw errors.badRequest(
      `The provided image content type is not supported. Supported content types are: ${IMAGE_CONTENT_TYPES.join(
        `, `,
      )}`,
    );
  }

  if (!isSupportedExtension(extension)) {
    throw errors.badRequest(
      `The provided image extension is not supported. Supported extensions are: ${IMAGE_EXTENSIONS.join(
        `, `,
      )}`,
    );
  }

  const SIZE_IN_MB_LIMIT = 4;

  if (toUnit(size, `mb`) > SIZE_IN_MB_LIMIT) {
    throw errors.badRequest(
      `Max image size is ${SIZE_IN_MB_LIMIT} megabytes (MB)`,
    );
  }

  const id = uuid();
  const location = `${uid}/images/${id}`;
  const bucket = storage().bucket();
  const file = bucket.file(location);
  const url = createUrl({ name: bucket.name, id, uid });

  const imagesModel: ImagesModel = {
    [id]: {
      extension,
      contentType,
      url,
    },
  };

  const userImagesRef = db.collection(`images`).doc(uid);

  const [userImagesSnap] = await Promise.all([
    userImagesRef.get(),
    file.save(buffer, {
      contentType,
    }),
  ]);

  const userImages = userImagesSnap.data();

  userImages
    ? await userImagesRef.update(imagesModel)
    : await userImagesRef.set(imagesModel);

  return {
    contentType,
    extension,
    id,
    url,
  };
};

export { uploadImageHandler };
