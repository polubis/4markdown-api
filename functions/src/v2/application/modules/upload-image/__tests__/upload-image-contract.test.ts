import { uploadImagePayloadSchema } from '../upload-image.contract';

describe(`Upload image contract works when`, () => {
  it(`accepts a valid base64 image data`, () => {
    const validBase64Image = `data:image/jpeg;base64,dGVzdGluZyBjb250ZW50`;

    expect(() =>
      uploadImagePayloadSchema.parse({
        image: validBase64Image,
      }),
    ).not.toThrowError();
  });

  it(`rejects partial base64 string without data URI prefix`, () => {
    const validBase64Image = `/9j/4AAQSkZJRgABAQEAAAAAA...`;

    const payload = {
      image: validBase64Image,
    };

    expect(() => uploadImagePayloadSchema.parse(payload)).toThrowError();
  });

  it(`rejects invalid base64 string`, () => {
    const invalidBase64Image = `invalid-base64-string`;

    const payload = {
      image: invalidBase64Image,
    };

    expect(() => uploadImagePayloadSchema.parse(payload)).toThrowError();
  });

  it(`rejects payload without the image field`, () => {
    const payload = {};

    expect(() => uploadImagePayloadSchema.parse(payload)).toThrowError();
  });

  it(`rejects an empty image field`, () => {
    const payload = {
      image: ``,
    };

    expect(() => uploadImagePayloadSchema.parse(payload)).toThrowError(
      `String must contain at least 1 character(s)`,
    );
  });

  it(`rejects base64 data with invalid characters`, () => {
    const invalidBase64ImageWithChars = `data:image/jpeg;base64,abc!def==`;

    const payload = {
      image: invalidBase64ImageWithChars,
    };

    expect(() => uploadImagePayloadSchema.parse(payload)).toThrowError();
  });
});
