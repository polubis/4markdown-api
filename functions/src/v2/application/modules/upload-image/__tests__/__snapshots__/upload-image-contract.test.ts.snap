// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload image contract works when rejects an empty image field 1`] = `
{
  "error": [ZodError: [
  {
    "code": "too_small",
    "minimum": 1,
    "type": "string",
    "inclusive": true,
    "exact": false,
    "message": "Image format is not encoded in base64",
    "path": [
      "image"
    ]
  },
  {
    "validation": "regex",
    "code": "invalid_string",
    "message": "Image format is not encoded in base64",
    "path": [
      "image"
    ]
  }
]],
  "success": false,
}
`;
